lib/mocha.js

```diff
@@ -97,6 +101,7 @@ function Mocha(options) {
   this.options = options;
   // root suite
   this.suite = new exports.Suite('', new exports.Context(), true);
+  this._cleanReferencesAfterRun = true;
```

Comment:
do you think it'd be better to use a finite state machine here?  when we start dealing with more than a couple boolean flags that get mutated, it starts to get hairy...

Reply:
Yeah, although I think `_cleanReferencesAfterRun` should still be a separate  boolean, since it is independent. The other state bits can be converted I think. Let me see what I can come up with.

Reply:
I've implemented a simple state machine. Will push as soon as local build is green 💚

```js
/**
 * A Mocha instance is a finite state machine.
 * These are the states it can be in.
 */
var mochaStates = utils.defineConstants({
  /**
   * Initial state of the mocha instance
   */
  INIT: 'init',
  /**
   * Mocha instance is running tests
   */
  RUNNING: 'running',
  /**
   * Mocha instance is done running tests and references to test functions and hooks are cleaned.
   * You can reset this state by unloading the test files.
   */
  REFERENCES_CLEANED: 'referencesCleaned',
  /**
   * Mocha instance is disposed and can no longer be used.
   */
  DISPOSED: 'disposed'
});
```